File: splitsy-backend/src/domain/common/value_objects/id.py
from dataclasses import dataclass
from uuid import UUID, uuid4

from src.domain.common.errors import ValidationError
from src.domain.common.value_objects.value_object import ValueObject


@dataclass(frozen=True)
class IdProps:
    value: str


class Id(ValueObject[IdProps]):
    def __init__(self, props: IdProps) -> None:
        super().__init__(props)
        self._value = props.value

    @property
    def value(self) -> str:
        return self._value

    @staticmethod
    def create(id: str | None = None) -> "Id":
        if id is None:
            return Id(IdProps(value=str(uuid4())))

        try:
            uuid_obj = UUID(id)

            return Id(IdProps(value=str(uuid_obj)))
        except (AttributeError, ValueError):
            raise ValidationError(f"Invalid Id format: {id}")

    def to_uuid(self) -> UUID:
        return UUID(self.value)

    def __repr__(self) -> str:
        return f"Id(value='{self.value}')"

    def __str__(self) -> str:
        return self.value

File: splitsy-frontend/src/domain/shared/value-objects/Id.ts
import { ValueObject } from "./ValueObject";

export interface IdProps {
  value: string;
}

export class Id extends ValueObject<IdProps> {
  private constructor(props: IdProps) {
    super(props);
  }

  public static create(): Id {
    return new Id({ value: Math.random().toString(36).replace('0.', '') })
  }

  public static createFrom(value: string): Id {
    return new Id({ value })
  }

  public equals(id: Id): boolean {
    return this.value === id.value;
  }

  public get value(): string {
    return this.props.value;
  }
}